sdk/rust/nym-sdk/src/stream_wrapper/mixnet_stream_wrapper_ipr.rs

```diff
@@ -0,0 +1,777 @@
+
+        self.send_ipr_request(request).await?;
+        self.listen_for_connect_response(request_id).await
+    }
+
+    async fn listen_for_connect_response(&mut self, request_id: u64) -> Result<IpPair, Error> {
+        let timeout = tokio::time::sleep(IPR_CONNECT_TIMEOUT);
+        tokio::pin!(timeout);
+
+        let mut framed = FramedRead::new(&mut self.stream, ReconstructedMessageCodec {});
+
+        loop {
+            tokio::select! {
+                _ = &mut timeout => {
+                    return Err(Error::IPRConnectResponseTimeout);
+                }
+                frame = framed.next() => {
+                    match frame {
+                        None => {
+                            return Err(Error::IPRClientStreamClosed);
+                        }
+                        Some(Err(e)) => {
+                            return Err(Error::MessageRecovery(e));
+                        }
+                        Some(Ok(reconstructed)) => {
+                            if let Err(e) = check_ipr_message_version(&reconstructed) {
+                                error!("Version check failed: {}", e);
+                                continue;
+                            }
+                            if let Ok(response) = IpPacketResponse::from_reconstructed_message(&reconstructed) {
+                                if response.id() == Some(request_id) {
+                                    return self.handle_connect_response(response).await;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    async fn handle_connect_response(&self, response: IpPacketResponse) -> Result<IpPair, Error> {
+        let control_response = match response.data {
+            IpPacketResponseData::Control(c) => c,
+            other => return Err(Error::UnexpectedResponseType(other)),
+        };
+
+        match *control_response {
+            ControlResponse::Connect(connect_resp) => match connect_resp.reply {
+                ConnectResponseReply::Success(success) => Ok(success.ips),
+                ConnectResponseReply::Failure(reason) => Err(Error::ConnectDenied(reason)),
+            },
+            _ => Err(Error::UnexpectedResponseType(
+                IpPacketResponseData::Control(control_response.clone()),
+            )),
+        }
+    }
+
+    pub async fn send_ip_packet(&mut self, packet: &[u8]) -> Result<(), Error> {
+        if self.connection_state != ConnectionState::Connected {
+            return Err(Error::IprStreamClientNotConnected);
+        }
+        let request = IpPacketRequest::new_data_request(packet.to_vec().into());
+        self.send_ipr_request(request).await
+    }
+
+    pub async fn handle_incoming(&mut self) -> Result<Vec<Bytes>, Error> {
+        let mut framed = FramedRead::new(&mut self.stream, ReconstructedMessageCodec {});
```

Comment:
it seems that whenever we read from the stream we always wrap it in the codec. couldn't we instead store the wrapped stream on `self` directly instead?